itou/openid_connect/pe_connect/views.py

```diff
@@ -0,0 +1,195 @@
+        token_data = response.json()
+    else:
+        token_data = dict(parse_qsl(response.text))
+
+    if not token_data or "access_token" not in token_data:
+        error_msg = "Aucun champ « access_token » dans la réponse PôleEmploiConnect, impossible de vous authentifier"
+        return _redirect_to_job_seeker_login_on_error(error_msg, request)
+
+    access_token = token_data.get("access_token")
+
+    # A token has been provided so it's time to fetch associated user infos
+    # because the token is only valid for 5 seconds.
+    url = constants.PE_CONNECT_ENDPOINT_USERINFO
+
+    response = httpx.get(
+        url,
+        params={"schema": "openid"},
+        headers={"Authorization": "Bearer " + access_token},
+        timeout=60,
+    )
+    if response.status_code != 200:
+        error_msg = "Impossible d'obtenir les informations utilisateur de PôleEmploiConnect."
+        return _redirect_to_job_seeker_login_on_error(error_msg)
+
+    try:
+        user_data = json.loads(response.content.decode("utf-8"))
+    except json.decoder.JSONDecodeError:
+        error_msg = "Impossible de décoder les informations utilisateur."
+        return _redirect_to_job_seeker_login_on_error(error_msg)
+
+    if "sub" not in user_data:
+        # 'sub' is the unique identifier from PôleEmploiConnect, we need that to match a user later on
+        error_msg = "Le paramètre « sub » n'a pas été retourné par PôleEmploiConnect. Il est nécessaire pour identifier un utilisateur."  # noqa E501
+        logger.error(error_msg)
+        return _redirect_to_job_seeker_login_on_error(error_msg)
+
+    pe_user_data = PoleEmploiConnectUserData.from_user_info(user_data)
+
+    try:
+        # At this step, we can update the user's fields in DB and create a session if required
+        user, _ = pe_user_data.create_or_update_user()
+    except InvalidKindException as e:
+        messages.info(request, "Ce compte existe déjà, veuillez vous connecter.")
+        url = {
+            UserKind.PRESCRIBER: reverse("login:prescriber"),
+            UserKind.SIAE_STAFF: reverse("login:siae_staff"),
+            UserKind.LABOR_INSPECTOR: reverse("login:labor_inspector"),
+        }[e.user.kind]
+        return HttpResponseRedirect(url)
+    except MultipleUsersFoundException as e:
+        return _redirect_to_job_seeker_login_on_error(
+            "Vous avez deux comptes sur la plateforme et nous détectons un conflit d'email : "
+            f"{e.users[0].email} et {e.users[1].email}. "
+            "Veuillez vous rapprocher du support pour débloquer la situation en suivant "
+            f"<a href='{global_constants.ITOU_HELP_CENTER_URL}'>ce lien</a>.",
+            request=request,
+        )
+
+    nir = request.session.get(global_constants.ITOU_SESSION_NIR_KEY)
+    if nir:
+        user.nir = nir
+        user.lack_of_nir_reason = ""
+        user.save(update_fields=["nir", "lack_of_nir_reason"])
```

Comment:
not bad

Reply:
:+1:

Il faudrait peut-être gérer le cas d’un conflit de NIR (compte dupliqué) ?

Reply:
En fait, la vue qui s'occupe de placer le nir en session vérifie qu'il est unique, du coup ça ne devrait pas arriver (sauf tentatives simultanées de création de compte, ce qui est peu probable)